SAK-46810 Adding buttons and confirmation in Grader to avoid that user goes out without confirm unsaved changes

feralvarez95 opened the pull request:
https://sakaiproject.atlassian.net/browse/SAK-46810
- Added the button for feedback and private notes.
- Added confirmation in both cases if the user try to close the dialog without save a modified text.
- Added confirmation in this two cases. Before this PR, this allows user return to student list with unsaved changes in this two fields without confirmation.

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -579,53 +595,134 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
   }
 
   doneWithPrivateNotesDialog() {
-
```
Can you leave these blank lines in please? They improve readability.

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -167,7 +169,7 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       </div>
       <div class="grader-navigator">
-        <div><a class="user-list-link" href="${this.userListUrl}" title="${this.assignmentsI18n["nav.list"]}">${this.assignmentsI18n["nav.list"]}</a></div>
+        <div><a class="user-list-link" href="${this.userListUrl}" title="${this.assignmentsI18n["nav.list"]}" @click=${this.toStudentList}>${this.assignmentsI18n["nav.list"]}</a></div>
```
I'm confused here. Why add a click handler when the href is already set? I know you must have a reason - I'm just asking.

feralvarez95 replied:
The aim is to not navigate when Feedback or Private Notes are changed but are not save. If I remove href, the link style is lost and it looks like normal text. Any suggestion to solve this in a better way would be great.

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -29,6 +29,8 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
     this.hasUngraded = false;
     this.hasUnsubmitted = false;
     this.resubmitNumber = "1";
+    this.previousPrivateNotes = "";
```
You shouldn't need to do this. We already store the original submissions and the Submission class has a privateNotes field. We need to try avoid duplicating data structures.

feralvarez95 replied:
I shaw that, you are right! That was not necesary! Thanks!

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -579,53 +595,134 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
   }
 
-  togglePrivateNotes() {
+  isFeedbackUnsaved() {
```
I don't think you should need this. Look at the way the modified (dirty) flag is throughout the code.

feralvarez95 replied:
Now I change to use dirty! I made some local tests, hope is fine now too.

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -579,53 +595,134 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+    return this.feedbackCommentEditor.getData() !== this.previousFeedback;
+  }
 
+  isPvtNotesUnsaved() {
```
Look at the modified flag usage that's already there. If you want to indicate that it is the private notes that has changed, maybe change the colour of the private notes button, or add an alert icon after it. Actually, the icon is better for a11y.

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -579,53 +595,134 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+    this.toggleFeedback();
+  }
+
+  confirmNotSavePvtNotes() {
+    this.submission.privateNotes = this.previousPrivateNotes;
```
```suggestion
  confirmNotSavePvtNotes() {
  
    this.submission.privateNotes = this.previousPrivateNotes;
```

feralvarez95 commented:
@adrianfish all changes done! Thanks for the review!!

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -598,22 +622,56 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+    }
+    else if (this.privateNotesEditor.checkDirty()) {
+      this.toggleConfirm("confirm-private-notes-panel");
+    }
+    else {
```
```suggestion
    } else {
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -622,10 +680,47 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+    if(!confirmPanel.dialog("instance")) {
+      confirmPanel.dialog({
+        width: "auto",
+        close: () => this.toggleConfirm(panelId)
```
```suggestion
        close: () => this.toggleConfirm(panelId),
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -622,10 +680,47 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+        width: "auto",
+        close: () => this.toggleConfirm(panelId)
+      });
+    }
+    else {
```
```suggestion
    } else {
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -881,6 +976,13 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       : true;
   }
 
+  toStudentList(e){
+    e.preventDefault();
```
```suggestion
  toStudentList(e) {
  
    e.preventDefault();
```

adrianfish reviewed (changes requested):
some more small things. Thanks for making the changes. I'm going to mark this as future just to make sure it doesn't get merged before 22.0 is out. Nothing bad! I have loads of stuff waiting for 22.0 as well. If we merge before and a fix goes into the same area of code there's a risk we'll backmerge parts of this work.

adrianfish commented:
@feralvarez95 I see it's marked a 22 issue, so no future label needed :)

adrianfish commented:
Could you post a couple of screenshots of the results of your work?

jesusmmp commented:
Yes @adrianfish, this is for 22 because fix some issues that were introduced in https://sakaiproject.atlassian.net/browse/SAK-45987

SedueRey commented:
@adrianfish I think we should add a linter to sakai to avoid all this spaces and style review. It's great you have noticed this but I think this should be made automatically.

SedueRey commented:
@adrianfish me too, my coworkers hate me :D, I like airbnb ones but any linter could be great as long as it's the same for everyone.

adrianfish commented:
We already have a linter in master although the keyword-spacing rules hasn't been added. I'll add that.

SedueRey commented:
@adrianfish A-W-E-S-O-M-E

jesusmmp commented:
> We already have a linter in master although the keyword-spacing rules hasn't been added. I'll add that.

Yes, please!

bjones86 commented:
So, there are problems with having a linter do things like this in the Sakai code base. First off, we don't have an accepted code style. There are files with spaces, files with tabs, files with crazy mixes of both. How will a linter decide which is appropriate? As far as I'm aware, linters only have one configuration, and it applies this to all files.

We also can't just go and change the entirety of the code base to use one style, because that would muddy up the history to the extent where it may become difficult to figure out when/who introduced specific lines of code.

This has all been brought up multiple times in the past. Unfortunately, we can never as a community agree on one specific coding style. Also unfortunate, we missed our opportunity to make sweeping, broad changes to the entire code base when we moved from SVN to Git.

Due to this, we are stuck in the pattern of manually reviewing patches for appropriate whitespace changes. At the current time, as it has been for all of Sakai's history, it is the developer's responsibility to stick to the coding style already present in a given file.

ottenhoff commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -167,7 +167,7 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       </div>
       <div class="grader-navigator">
-        <div><a class="user-list-link" href="${this.userListUrl}" title="${this.assignmentsI18n["nav.list"]}">${this.assignmentsI18n["nav.list"]}</a></div>
+        <div><a class="user-list-link" href="javascript:void(0)" title="${this.assignmentsI18n["nav.list"]}" @click=${this.toStudentList}>${this.assignmentsI18n["nav.list"]}</a></div>
```
i always prefer a -button- when i see javascript:void

feralvarez95 replied:
Currently this is showing a regular link (underlining blue text), our btn-link css classes shows it as normal text. Probably change this link and putting a normal button could be a good idea, because old grader shows a normal button.

feralvarez95 replied:
Finally I changed the link by a button, with the css now it looks like a button. How do yo see it @adrianfish @ottenhoff? 
![link-to-button](https://user-images.githubusercontent.com/19774177/156521343-26d3a32d-3d41-4fe4-b202-6e1b70da6a01.png)

adrianfish replied:
> i always prefer a -button- when i see javascript:void

@ottenhoff  But this link causes a navigation, not a popup. Aren't links fine in this situation?

ottenhoff replied:
Ok here is a good rule of thumb:

"Buttons are used for actions that affect the website’s front-end or back-end; links are used for navigation and actions that don’t affect the website at all."

https://ux.iu.edu/writings/buttons-vs-links-basic/#:~:text=There%20are%20differences%20as%20to,affect%20the%20website%20at%20all.

I guess I get suspicious when I see javascript in a link as it usually indicates that *navigation* is *not* being performed.

adrianfish commented:
@bjones86 I think we have to lint our JS as it makes it more reliable, catching lots of stuff that a compiler would catch. It's helped me so many times, catching my bad stuff :)

bjones86 commented:
I'm not saying linters are bad, they're just not a solution for the tabs vs spaces issue across the board in Sakai code. This is due to our own practises and choices, unfortunately.

adrianfish commented:
We're only using one in the webcomponents project, which is pretty much greenfield code. Elsewhere, I know exactly what you mean.

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -585,45 +598,101 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       feedbackPanel.dialog({
         width: "auto",
-        close: () => this.toggleFeedback()
+        beforeClose: () => { return this.cancelFeedbackToggle() },
```
You don't need the return here. () => this.cancelFeedbackToggle(),

feralvarez95 replied:
I think that I need to prevent close before confirm if we want to close. I don't know if there is a better approach.

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -585,45 +598,101 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       this.feedbackCommentEditor = this.replaceWithEditor("grader-feedback-comment");
-    } else {
-      this.submission.feedbackComment = this.feedbackCommentEditor.getData();
+    }
+    else {
```
```suggestion
    } else {
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -585,45 +598,101 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+        this.feedbackCommentEditor.resetDirty();
+        this.confirmedNotSaveFeedback = false;
+        document.getElementById("feedback-comment-unsaved-msg").classList.add("hidden");
+      }
+      else {
```
```suggestion
      } else {
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -585,45 +598,101 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+  cancelFeedbackToggle() {
+
+    if (this.feedbackCommentEditor.checkDirty()) {
+      if (this.confirmedNotSaveFeedback){
```
```suggestion
      if (this.confirmedNotSaveFeedback) {
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -585,45 +598,101 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
+        this.privateNotesEditor.resetDirty();
+        this.confirmedNotSavePvtNotes = false;
+        document.getElementById("private-notes-unsaved-msg").classList.add("hidden");
+      }
+      else {
```
```suggestion
      } else {
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -585,45 +598,101 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       privateNotesPanel.dialog({
         width: "auto",
-        close: () => this.togglePrivateNotes()
+        beforeClose: () => { return this.cancelPrivateNotesToggle() },
```
```suggestion
        beforeClose: () => this.cancelPrivateNotesToggle(),
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -585,45 +598,101 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       this.privateNotesEditor = this.replaceWithEditor("grader-private-notes");
-    } else {
-      this.submission.privateNotes = this.privateNotesEditor.getData();
+    }
+    else {
```
```suggestion
    } else {
```

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -781,6 +852,8 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       const rubricGrading = document.getElementsByTagName("sakai-rubric-grading").item(0);
       rubricGrading && rubricGrading.release();
+      document.getElementById("feedback-active-indicator").classList.remove("unsaved");
+      document.getElementById("pvt-notes-active-indicator").classList.remove("unsaved");
```
Are you sure these elements are always present?

feralvarez95 replied:
I changed this by two properties in the component, now is updated when the properties are updated (I think it is much better now)

adrianfish commented on webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:
```diff
@@ -881,6 +954,14 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
       : true;
   }
 
+  toStudentList(e){
```
```suggestion
  toStudentList(e) {
```

adrianfish reviewed (dismissed):
Just small bits. I've added commit suggestions for most of them. Are you happy with the indicator approach?

feralvarez95 commented:
> Just small bits. I've added commit suggestions for most of them. Are you happy with the indicator approach?

I'm happy with this approach, I don't know if it is the better color but I think that interface is lighter avoiding unnecesary dialogs.
Also I added different titles for the case on active-indicator is unsaved.
